### PR TITLE
[ASR] Add optimization util for linear sum assignment algorithm

### PR DESCRIPTION
**Pull request by tango4j**
# What does this PR do ?

1. Add a optimization util file for linear sum assignment (LSA) solver algorithm for online-diarization/multi-speaker-ASR
LSA problem solver is needed for the following tasks in NeMo: 
        (1) Permutation Invariant Loss (PIL) for diarization model training
        (2) Label permutation matching for online speaker diarzation 
        (3) Concatenated minimum-permutation Word Error Rate (cp-WER) calculation

What is LSA solver algorithm? [Google OR-tools LSA Solver](https://developers.google.com/optimization/assignment/linear_assignment)

The NeMo linear_sum_assignment function is compared with `scipy.optimization.linear_sum_assingment`.
In the [unit-test ](https://github.com/NVIDIA/NeMo/blob/49d4e80ddf1138cea9593392a9a9e76be9ca58ec/tests/collections/asr/test_diar_utils.py#L792)for NeMo LSA solver, the result is compared with the `scipy` version of `linear_sum_assignment`. 

2. Removing `@torch.jit.script` decorator in speaker_utils.py since it creates type-errors when the code is not used for production purpose. 
     Instead, all `torch.jit.script` required classes and functions are tested in `test_diar_utils.py`. 
     Take a look at [these tests](https://github.com/NVIDIA/NeMo/blob/49d4e80ddf1138cea9593392a9a9e76be9ca58ec/tests/collections/asr/test_diar_utils.py#L606) for checking jit_script = [True/False] and cuda = [True/False] (testing total 4 combinations)

3. Also refactored some of the functions in online diarization 
    - replaced scipy LSA solver to NeMo LSA solver in `online_clustering.py`.

5. Added a couple of functions in `der.py` for online diarization DER calculation. 
    - replaced scipy LSA solver to NeMo LSA solver in `der.py`.

**Collection**: [ASR]

# Changelog 
- nemo/collections/asr/metrics/der.py
: replaced scipy LSA solver to NeMo LSA solver in `calculate_session_cpWER ` function. 
: Added two functions for online diarization evaluations: `get_partial_ref_labels` and `get_online_DER_stats`.

- nemo/collections/asr/models/online_diarizer.py
: Made `_perform_online_clustering` function simpler by moving `get_reduced_mat` and `match_labels` into online clustering function. 

- nemo/collections/asr/parts/utils/offline_clustering.py
: Added `laplacian = laplacian.float().to(torch.device('cpu'))` to avoid jit-scripted module uses GPU even when CPU is specified or vice-versa. This behavior is always tested/checked in `test_diar_utils.py`.

- nemo/collections/asr/parts/utils/online_clustering.py
 : replaced scipy LSA solver to NeMo LSA solver in `get_lsa_speaker_mapping` function.
 : Modified the docstrings of `update_speaker_history_buffer` to make the example easier.

- nemo/collections/asr/parts/utils/optimization_utils.py
  : Fully torch-jit-scriptable, linear sum assignment problem solver class and function were added.

- nemo/collections/asr/parts/utils/speaker_utils.py
 : Removed `@torch.jit.script` decorators since this creates unnecessary warning messages and type related errors when used without scripting.

- tests/collections/asr/test_diar_metrics.py
: Added unit-tests for the newly added function `get_partial_ref_labels` and `get_online_DER_stats`.

- tests/collections/asr/test_diar_utils.py
: Added tests for offline clustering and online clustering for many different cases including:
[jit-script=True, cuda=True], 
[jit-script=True, cuda=False], 
[jit-script=False, cuda=True], 
[jit-script=False, cuda=False] cases 
 which is using the torch-jit-scripted NeMo linear_sum_assignment function.

# Usage

```python
from nemo.collections.asr.parts.utils.optimization_utils import linear_sum_assignment
#An example cost matrix to be solved
cost_matrix = \
torch.tensor([[7, 6, 2], [6, 2, 1], [5, 6, 8]])
row_ind_nm, col_ind_nm = linear_sum_assignment(cost_matrix)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.

### CODE REVIEW

File: nemo/collections/asr/metrics/der.py [hunk 1]
```diff
@@ -32,6 +34,77 @@
...
+    last_pred_time = float(pred_labels[-1].split()[1])
+    ref_labels_out = []
+    for label in ref_labels:
+        start, end, speaker = label.split()
+        start, end = float(start), float(end)
+        # If ref start time passes the last prediction time, break
+        if start >= last_pred_time:
+            break
+        # If the current [start, end] interval is latching the last prediction time
+        if start < last_pred_time <= end:
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/metrics/der.py]: Foxed

File: nemo/collections/asr/models/online_diarizer.py [hunk 2]
```diff
@@ -431,10 +431,16 @@
...
         if end_idx > stt_idx:
             torch_embs = self._run_embedding_extractor(audio_signal[stt_idx:end_idx])
-            if embeddings is None:
+            if embeddings is None or embeddings.shape[0] == 0:
                 embeddings = torch_embs
             else:
-                embeddings = torch.vstack((embeddings[:stt_idx, :], torch_embs))
+                try:
+                    embeddings = torch.vstack((embeddings[:stt_idx, :], torch_embs))
+                except:
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py]: Fixed

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 3]
```diff
@@ -552,12 +550,13 @@ def eigDecompose(
             device = torch.cuda.current_device()
         laplacian = laplacian.float().to(device)
     else:
+        laplacian = laplacian.float().to(torch.device('cpu'))
         laplacian = laplacian.float()
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/offline_clustering.py:554]: why same operation twice?

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 4]
```diff
@@ -552,12 +550,13 @@ def eigDecompose(
...
     lambdas, diffusion_map = eigh(laplacian)
     return lambdas, diffusion_map
 
 
-def eigValueSh(laplacian: torch.Tensor, cuda: bool, device: torch.device = torch.device('cpu')) -> torch.Tensor:
+def eigValueSh(laplacian: torch.Tensor, cuda: bool, device: torch.device) -> torch.Tensor:
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/offline_clustering.py:559]: why cuda and device? Isn't only one sufficient

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 5]
```diff
@@ -566,6 +565,7 @@ def eigValueSh(laplacian: torch.Tensor, cuda: bool, device: torch.device = torch
             device = torch.cuda.current_device()
         laplacian = laplacian.float().to(device)
     else:
+        laplacian = laplacian.float().to(torch.device('cpu'))
         laplacian = laplacian.float()
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/offline_clustering.py:569]: same here. laplacian.float() twice

File: nemo/collections/asr/parts/utils/online_clustering.py [hunk 6]
```diff
@@ -35,63 +35,63 @@
...
     """
-    enc = OneHotEncoder(handle_unknown='ignore')
     all_spks_labels = [[x] for x in range(len(U_set))]
-    enc.fit(all_spks_labels)
-    enc_P = enc.transform(cmm_P.reshape(-1, 1)).toarray()
-    enc_Q = enc.transform(cmm_Q.reshape(-1, 1)).toarray()
-    stacked = np.hstack((enc_P, enc_Q))
-    cost = -1 * linear_kernel(stacked.T)[spk_count:, :spk_count]
-    row_ind, col_ind = linear_sum_assignment(cost)
+    PandQ_list: List[int] = [int(x.item()) for x in PandQ]
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/online_clustering.py:70]: minor: mentioning dtype in variable name need to be avoided

**Review by nithinraok — COMMENTED** [review]
minor review. Will do thorough review tomorrow. 
Very neat improvement, need to understand better from my end.

File: nemo/collections/asr/metrics/der.py [hunk 7]
```diff
@@ -32,6 +34,83 @@
...
+    if len(pred_labels) == 0:
+        last_pred_time = 0
+    else:
+        # The lastest prediction time in the prediction labels
+        last_pred_time = max([float(labels.split()[1]) for labels in pred_labels])
+    ref_labels_out = []
+    for label in ref_labels:
+        start, end, speaker = label.split()
+        start, end = float(start), float(end)
+        # If the current [start, end] interval is latching the last prediction time
```
Comment (nithinraok) [inline nemo/collections/asr/metrics/der.py:63]: latching -> matching

Comment (tango4j) [reply nemo/collections/asr/metrics/der.py:63]: Changed the expression (Checked by Elena)

File: nemo/collections/asr/parts/utils/offline_clustering.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:554]: removed

File: nemo/collections/asr/parts/utils/offline_clustering.py (hunk 5 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:569]: removed

File: nemo/collections/asr/parts/utils/optimization_utils.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+# The following is the full text of the license:
+
+# Hungarian algorithm (Kuhn-Munkres) for solving the linear sum assignment
+# problem. Taken from scikit-learn. Based on original code by Brian Clapper,
+# adapted to NumPy by Gael Varoquaux.
+# Further improvements by Ben Root, Vlad Niculae and Lars Buitinck.
+# Copyright (c) 2008 Brian M. Clapper <bmc@clapper.org>, Gael Varoquaux
+# Author: Brian M. Clapper, Gael Varoquaux
+# License: 3-clause BSD
+
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/optimization_utils.py:26]: I think we can add other algorithms below this.
The copyright in the beginning of the code is the convention in the most of the project so I followed

File: nemo/collections/asr/parts/utils/optimization_utils.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        cost_mat (Tensor): 2D matrix containing cost matrix. Number of columns must be larger than number of rows.
+        row_uncovered (Tensor): 1D matrix containing boolean values indicating whether a row is covered.
+        col_uncovered (Tensor): 1D matrix containing boolean values indicating whether a column is covered.
+        zero_row (Tensor): 1D matrix containing the row index of the last zero found.
+        zero_col (Tensor): 1D matrix containing the column index of the last zero found.
+        path (Tensor): 2D matrix containing the path taken through the matrix.
+        marked (Tensor): 2D matrix containing the marked zeros.
+    """
+
+    def __init__(self, cost_matrix):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/optimization_utils.py:95]: If there is no type annotation, jit compiler think of it as torch.Tensor.

File: nemo/collections/asr/parts/utils/optimization_utils.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+        raise ValueError(
+            f"Cost matrix size {cost_matrix.shape} is too large. The maximum supported size is {max_size}x{max_size}."
+        )
+
+    # The algorithm expects more columns than rows in the cost matrix.
+    if cost_matrix.shape[1] < cost_matrix.shape[0]:
+        cost_matrix = cost_matrix.T
+        transposed = True
+    else:
+        transposed = False
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/optimization_utils.py:316]: This followed the original implementation in scipy.
If we don't use `transposed` variable, we need to create another variable to indicate that `foo = cost_matrix.shape[1] < cost_matrix.shape[0]`.

File: nemo/collections/asr/parts/utils/offline_clustering.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py:559]: This was added long back because there are users setting cuda=True but device=cpu.
This is adding some flexibility to avoid errors on such cases.
If we need to remove this, lt requires a speparate PR since this involves whole diarization pipeline.

File: nemo/collections/asr/parts/utils/online_clustering.py (hunk 6 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/online_clustering.py:70]: Makes sense since types are strictly annotated for jit script functions.
Fixed.

File: nemo/collections/asr/parts/utils/optimization_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/optimization_utils.py]: Got it

**Review by nithinraok — APPROVED** [review]
LGTM
